### ISSUE TITLE
crev for android apps / binary artifacts

### ISSUE THREAD

**Issue by tiziano88** [crev-dev/crev#9]
I have been thinking about a similar mechanism to crev, but to determine the trustworthiness of executable binaries (e.g. android apps, native applications, JS code, etc.). I didn't know about crev when I started thinking about it, but now that I found out about it, I think it could leverage the same web of trust as crev, and ideally also the same format for proofs (using hashes to identify specific artifacts).

From a user's point of view, there may be an agent installed their device that scans binaries and apps and compares them against reviewed apps from their web of trust. One way to think about it would be something like an outsourced version of https://f-droid.org/ , but cross-platform and with "pluggable trust".

In order for this to work, it would be necessary (but not sufficient) to have reproducible builds for the binaries in question (so that different verifiers / reviewers can independently build a bitwise identical artifact starting from the same source).

We may need different roles in the verification process, for instance a builder that is only trusted to confirm that a specific artifact can be traced back to some source code, which is separately reviewed by actual human reviewers, and their reviews should transfer over from the snapshot of the source code, to the computed binary.

An example of a problem that this could solve is similar to what described in https://core.telegram.org/reproducible-builds , but in a more automated way.

Is there interested in considering something like this as an extension of crev? Has anyone already thought about it?

**Comment by dpc** [comment]
That's not the first time I'm hearing about binaries signing, and it's in a way it's already done one way or another in different areas. AFAIR there are existing projects and companies doing things like this. Unfortunately my Google-fu and memory are failing me right now.

The biggest difference with binaries, is that they are not really reviewable. One can just attest that they authentic, but there's little room for someone to jump in and "review" anything on their own. This makes it less like that the WoT will be useful. 

Only Google can attest that their Chrome build is authentic, so why bother with WoT?

AFAIK, not a lot of project has reproducible builds, and the ones that do are usually so important that if the build is reproducible one can just build it themselves in the first place.

I'm not trying to be negative - just wondering if this would gain a lot of traction.

Now, having said that, it's is possible to reuse crev's proof format. It's been designed to be arbitrarily extensible, so technically it works just fine, along with WoT, etc.

**Comment by tiziano88** [comment]
I agree with the premises, but I actually think there is a lot that could be improved from the current state of things of binary supply chain.

As you mentioned, currently we blindly trust the signature (if at all) attached to the binary itself, to signify that the binary was packaged by the original author. This skips an important step of the software supply chain, namely from sources to binary artifacts.

Let's pick Chromium instead of Chrome as an example (since AFAICT Chrome is not actually open source). If Chromium were reproducibly buildable (which I am not sure it actually is), it should be possible for trusted reviewers to publish a proof that binary with hash H can be reproducibly built from the source code in the given repository at commit C.

This would ensure that

- the packager of the final binary may not have a chance to introduce backdoors / modifications to the binary itself before publishing it publicly
- the owner of the signing key may not use the key to sign a separate binary for a specific user (targeted backdoor)

Perhaps this is not super relevant for high profile projects like Chrome / Chromium, in which it is probably fine to trust Google directly for both authoring the code and packaging the binary, but say, I want to download a new fancy text editor from some random GitHub project; I go straight to the release page and download one of the available binaries. It would be nice if I could check that a number of people actually verified that such binary (identified by its hash) was indeed compiled directly from the checked-in sources.

Additionally, the sources themselves may or may not be verified by crev; if they are, then I would get even more information whether or not I should trust the functionality as defined in the source code, based on other people's reviews (and not only for the immediate source, but also for its dependencies); if not, I at least know that if I at a later point I (or someone else) wanted to review the code, I would be able to pull it form GitHub (modulo the author destroying the code by force pushing to the repository, which anyways would be pretty obvious).

**Comment by dpc** [comment]
I kind of agree. But the root of the problem is ... who's going to force people to have reproducible builds? 

Anyway. It's a worthy cause. :)

**Comment by tiziano88** [comment]
Yeah of course reproducible builds are necessary (but not sufficient), but more and more projects and build systems are moving towards that already. While we live in a world in which the same source code can be compiled down to a different binary based on who / when compiles it, then I would say we are in an even worse situation than the one that crev is trying to solve for source code review :) Imagine 10 people (claim to) compile the same source of Chromium and end up with different binaries, with no way of reconciling them to each other; why would you trust any of them to not have introduced a backdoor in the binary before serving it to you?

Anyways, the good news is that I think the proof format for crev is already extensible enough that this should be easy to implement on top of it.

In a sense, a binary artifact is not too dissimilar from a Rust crate made of a giant inline assembly blob and no external dependencies, from crev's point of view :)

**Comment by dpc** [comment]
> Anyways, the good news is that I think the proof format for crev is already extensible enough that this should be easy to implement on top of it.

Yeap. There should be no problems there. Everything was made so it's easy to reuse for practically whatever.

I wouldn't be surprised that **if** crev become successful, it's WoT could actually overtake PGP, and reused in completely unrelated applications that benefit from a WoT. One can put anything in these yamls, and sign it, simple yet powerful idea and I'm surprised that it's not more popular.
